src/fdc3/dotnet/DesktopAgent.Client/src/MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client/Infrastructure/Internal/Protocol/PrivateChannel.cs

```diff
@@ -0,0 +1,609 @@
+﻿/*
+ * Morgan Stanley makes this available to you under the Apache License,
+ * Version 2.0 (the "License"). You may obtain a copy of the License at
+ *
+ *      http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * See the NOTICE file distributed with this work for additional information
+ * regarding copyright ownership. Unless required by applicable law or agreed
+ * to in writing, software distributed under the License is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express
+ * or implied. See the License for the specific language governing permissions
+ * and limitations under the License.
+ */
+
+using System.Text.Json;
+using Finos.Fdc3;
+using Finos.Fdc3.Context;
+using Microsoft.Extensions.Logging;
+using Microsoft.Extensions.Logging.Abstractions;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Shared;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Shared.Contracts;
+using MorganStanley.ComposeUI.Fdc3.DesktopAgent.Shared.Exceptions;
+using MorganStanley.ComposeUI.Messaging.Abstractions;
+using DisplayMetadata = MorganStanley.ComposeUI.Fdc3.DesktopAgent.Shared.Protocol.DisplayMetadata;
+
+namespace MorganStanley.ComposeUI.Fdc3.DesktopAgent.Client.Infrastructure.Internal.Protocol;
+
+internal class PrivateChannel : Channel, IPrivateChannel, IAsyncDisposable
+{
+    private readonly ILogger<PrivateChannel> _logger;
+    private readonly ILoggerFactory _loggerFactory;
+    private readonly IMessaging _messaging;
+    private readonly PrivateChannelTopics _privateChannelTopics;
+    private readonly string _channelId;
+    private readonly string _instanceId;
+    private readonly string _internalEventsTopic;
+    private readonly bool _isOriginalCreator;
+    private readonly string _remoteContextListenersService;
+    private readonly TaskCompletionSource<string> _initializationTaskCompletionSource = new(TaskCreationOptions.RunContinuationsAsynchronously);
+
+    private readonly SemaphoreSlim _contextHandlersLock = new(1, 1);
```

Comment:
Having 4 different semaphores seems like a bit of an overkill. I would consider streamlining them - either by merging them or just using Interlocked for entry checks.